frontend/src/main/scala/bloop/engine/Interpreter.scala

```diff
@@ -110,20 +110,27 @@ object Interpreter {
       f: State => Task[State]
   ): Task[State] = {
     val reachable = Dag.dfs(getProjectsDag(projects, state))
-    val allSources = reachable.iterator.flatMap(_.sources.toList).map(_.underlying)
-    val watcher = SourceWatcher(projects.map(_.name), allSources.toList, state.logger)
-    val fg = (state: State) => {
-      val newState = State.stateCache.getUpdatedStateFrom(state).getOrElse(state)
-      f(newState).map { state =>
-        watcher.notifyWatch()
-        State.stateCache.updateBuild(state)
-      }
-    }
+    val projectsSourcesAndDirs = reachable.map(_.allSourceFilesAndDirectories)
```

Comment:
Is it intentional that we're only resolving sources once, before file watching?

Reply:
Yes. New files created after the file watcher has started will still be picked up and watched.

Reply:
Are you sure about that? What we're doing here is resolving sources at the beginning of source watching, but the watch iterations don't force a source resolution?

Reply:
That's correct. There's a single call (per project) to `allSourceFilesAndDirectories`, and new files are picked up and watched.

Reply:
I'm not sure we're talking about the same thing. If I file watch, but then I create a new file that would pick up by a regex, that would not be caught by this implementation. Is that right?

Reply:
No, the new file will be watched. It'd be interesting to verify that's true on all platforms actually. It's probably watched only because you can only watch directories on OSX, or some detail like this 🤔

Reply:
As discussed offline, I think there are some cases where the file changes will not be picked up (directory hosting the regex not existing before the source resolution and with regexes that are more than two level deep) but I think handling those cases will become more complicated, so I'm fine merging as is, just wanted to point out some limitations with this approach. :+1:

Reply:
I think it should be fine to recursively watch `SourcesGlobs.directory` instead of expanding globs and watching the individual files. This would correctly catch new files in nested directories. There's a risk we would watch too large directories, but that's already the case with normal source directories.